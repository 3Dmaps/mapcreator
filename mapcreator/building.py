import re
import subprocess
import shutil
from os import path, listdir, makedirs, rename, remove, devnull
from io import StringIO
from zipfile import ZipFile, ZIP_DEFLATED
from mapcreator import persistence

BUILD_DIR = path.join(persistence.STATE_DIR, 'build')
FINALIZED_DIR = path.join(BUILD_DIR, 'finalized')

OUTPUT_FILENAME_REGEX = re.compile(r'^\[(\d+)\](.*)')
OUTPUT_FILENAME_FORMAT = '[{}]{}'

INTERNAL_FORMAT = 'GTiff'
INTERNAL_FILE_EXTENSION = 'tiff'

PROJECTION_IDENTIFIER = 'EPSG:3857'
OUTPUT_CELLSIZE = 10

OUTPUT_FORMAT = 'ENVI'
OUTPUT_FILE_EXTENSION = 'bin'
METADATA_FILE_EXTENSION = 'hdr'

FINAL_FILENAME_FORMAT = 'heightfile{}.' + OUTPUT_FILE_EXTENSION
FINAL_METADATA_FORMAT = 'heightfile{}.' + METADATA_FILE_EXTENSION

class BuildStatus:
    def __init__(self, index, heightfile, state):
        self.index = index
        self.output = StringIO()
        self.original_file = heightfile
        self.current_file = heightfile
        self.state = state
        self.result_files = []
    def add_result_file(self, file):
        self.result_files.append(file)
    def __str__(self):
        lines = ['Build results for {}:'.format(path.basename(self.original_file))]
        if self.result_files:
            lines.append('-Files created: {}'.format(', '.join(map(path.basename, self.result_files))))
        else:
            lines.append('-No files were created')
        if self.output.getvalue():
            lines.append('-Messages from GDAL:')
            lines.extend(self.output.getvalue().split('\n'))
        return '\n'.join(lines)

def call_command(command, buildstatus, debug = False):
    with subprocess.Popen(command, stdout=subprocess.PIPE, stderr=subprocess.PIPE) as process:
        if debug: buildstatus.output.write('\n[{}]\n'.format(command))
        stdout, stderr = process.communicate()
        if debug: buildstatus.output.write(stdout.decode('utf-8'))
        buildstatus.output.write(stderr.decode('utf-8'))


def init_build():
    if temp_build_files_exist():
        cleanup()
    makedirs(BUILD_DIR)
    makedirs(FINALIZED_DIR)

def get_output_path(input_path, new_extension = ''):
    input_filename = path.basename(input_path)
    match = OUTPUT_FILENAME_REGEX.match(input_filename)
    if match:
        next_index = int(match.group(1)) + 1
        new_filename = OUTPUT_FILENAME_FORMAT.format(next_index, match.group(2))
    else:
        new_filename = OUTPUT_FILENAME_FORMAT.format(0, input_filename)
    new_filename += ('.' + new_extension if new_extension else '')
    return path.join(BUILD_DIR, new_filename)

def prepare(buildstatus, debug = False):
    """
    First convert the file to GTiff (GDALs standard format) to make sure that everything works as expected
    """ 
    outpath = get_output_path(buildstatus.current_file, INTERNAL_FILE_EXTENSION)
    command = 'gdal_translate -of {} {} {}'.format(INTERNAL_FORMAT, buildstatus.current_file, outpath)
    call_command(command, buildstatus, debug)
    buildstatus.current_file = outpath

def cut_projection_window(buildstatus, debug = False):
    if buildstatus.state.has_window():
        outpath = get_output_path(buildstatus.current_file)
        command = 'gdalwarp -te {} {} {}'.format(buildstatus.state.get_window_string_lowerleft_topright(), buildstatus.current_file, outpath)
        call_command(command, buildstatus, debug)
        buildstatus.current_file = outpath

def reproject(buildstatus, debug = False):
    outpath = get_output_path(buildstatus.current_file)
    command = 'gdalwarp -tr {} {} -t_srs {} -r bilinear {} {}'.format(OUTPUT_CELLSIZE, OUTPUT_CELLSIZE, PROJECTION_IDENTIFIER, buildstatus.current_file, outpath)
    call_command(command, buildstatus, debug)
    buildstatus.current_file = outpath

def translate(buildstatus, debug = False):
    outpath = get_output_path(buildstatus.current_file, OUTPUT_FILE_EXTENSION)
    metapath = get_output_path(buildstatus.current_file, METADATA_FILE_EXTENSION)

    command = 'gdal_translate -of {} {} {}'.format(OUTPUT_FORMAT, buildstatus.current_file, outpath)
    call_command(command, buildstatus, debug)

    finalized_metaname = FINAL_METADATA_FORMAT.format(buildstatus.index)
    finalized_metapath = path.join(FINALIZED_DIR, finalized_metaname)
    rename(metapath, finalized_metapath)

    buildstatus.current_file = outpath
    buildstatus.add_result_file(finalized_metapath)


def finalize(buildstatus, debug = False):
    if buildstatus.current_file == buildstatus.original_file:
        raise RuntimeError('No actions have been done --> Not finalizing anything')
    final_filename = FINAL_FILENAME_FORMAT.format(buildstatus.index)
    final_path = path.join(FINALIZED_DIR, final_filename)
    rename(buildstatus.current_file, final_path)
    buildstatus.current_file = final_path
    buildstatus.add_result_file(final_path)

def package(package_name, files):
    with ZipFile(package_name, 'w', ZIP_DEFLATED) as package:
        for f in files:
            package.write(f, path.basename(f))

def temp_build_files_exist():
    return path.exists(BUILD_DIR)

def cleanup():
    shutil.rmtree(BUILD_DIR)

BUILD_ACTIONS = (
    prepare, cut_projection_window, reproject, translate, finalize
)