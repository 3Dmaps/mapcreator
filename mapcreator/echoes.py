import click

IDENTIFIER_SIZE = 7
PROGRESS_BAR_TEMPLATE = (' ' * (IDENTIFIER_SIZE + 1)) + '%(label)s [%(bar)s] %(info)s'

def format_echo(identifier, s):
    if identifier:
        format_string = "{:>" + str(IDENTIFIER_SIZE) + "}: {}"
    else:
        format_string = "{:" + str(IDENTIFIER_SIZE + 1) + "} {}"
    return format_string.format(identifier, s)

def info(s):
    click.echo(format_echo('', s))

def highlight(s):
    click.secho(format_echo('', s), fg='cyan')

def warn(s):
    click.secho(format_echo('WARNING', s), fg='yellow')

def error(s):
    click.secho(format_echo('ERROR', s), bg='red')

def success(s):
    click.secho(format_echo('SUCCESS', s), bg='green')

def prettyinput():
    return input(' '*(IDENTIFIER_SIZE + 2))
