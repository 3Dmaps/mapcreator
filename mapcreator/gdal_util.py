import json
import subprocess

class Gdalinfo:
    @classmethod
    def for_file(cls, path):
        with subprocess.Popen(['gdalinfo', '-json', path], stdout=subprocess.PIPE, stderr=subprocess.PIPE) as process:
            stdout, stderr = process.communicate()
            gdal_info_dict = json.loads(stdout.decode('utf-8'))
            gdal_info = Gdalinfo()
            gdal_info_coordinates = gdal_info_dict['wgs84Extent']['coordinates'][0]
            minX = gdal_info_coordinates[0][0]
            minY = gdal_info_coordinates[0][1]
            maxX = gdal_info_coordinates[0][0]
            maxY = gdal_info_coordinates[0][1]
            for coord in gdal_info_coordinates:
                minX = min(minX, coord[0])
                maxX = max(maxX, coord[0])
                minY = min(minY, coord[1])
                maxY = max(maxY, coord[1])
            
            gdal_info.minX = minX
            gdal_info.minY = minY
            gdal_info.maxX = maxX
            gdal_info.maxY = maxY

            return gdal_info
        

