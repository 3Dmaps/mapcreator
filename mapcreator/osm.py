from xml.etree import ElementTree

class OSMData:

    TAG_ROOT = 'osm'
    TAG_NODE = 'node'
    TAG_WAY = 'way'
    TAG_WAY_NODE = 'nd'
    TAG_TAG = 'tag'
    ATTRIB_ID = 'id'
    ATTRIB_REF = 'ref'
    ATTRIB_KEY = 'k'
    ATTRIB_VALUE = 'v'
    ATTRIB_LAT = 'lat'
    ATTRIB_LON = 'lon'
    KEY_LANDUSE = 'landuse'
    ACCEPTED_LANDUSES = ['meadow']
    KEY_HIGHWAY = 'highway'
    ACCEPTED_HIGHWAYS = ['footway', 'path']

    def __init__(self):
        self.node_filters = []
        self.way_filters = []

    @classmethod
    def load(cls, path):
        data = OSMData()
        data.tree = ElementTree.parse(path)
        data.preprocess()
        return data

    @classmethod
    def get_elem_id(cls, elem):
        return int(elem.get(OSMData.ATTRIB_ID))
    
    @classmethod
    def get_tag(cls, elem, key):
        for tagElement in elem.findall(OSMData.TAG_TAG):
            if tagElement.get(OSMData.ATTRIB_KEY) == key:
                return tagElement.get(OSMData.ATTRIB_VALUE)
        return None

    def preprocess(self):
        self.nodes = {}
        self.included_nodes = set()
        self.ways = {}
        self.included_ways = set()
        root = self.tree.getroot()
        if root.tag != OSMData.TAG_ROOT:
            raise ValueError('Invalid OSM XML Data - the root node\'s tag was not "osm"!')
        for child in root:
            if child.tag == OSMData.TAG_NODE:
                nodeid = OSMData.get_elem_id(child)
                self.nodes[nodeid] = child
                self.included_nodes.add(nodeid)
            elif child.tag == OSMData.TAG_WAY:
                wayid = OSMData.get_elem_id(child)
                self.ways[wayid] = child
                self.included_ways.add(wayid)

    def add_node_filter(self, *filters):
        """
        Adds filters. 
        Filters are functions that take a ElementTree element and this OSMData instance
        as arguments and returns a truthful value if said element should be
        kept.
        Filters given in the same method call are joined with "and",
        while filters given in subsequent calls are joined with "or".
        For example

        data = OSMData()
        data.add_node_filter(f1, f2, f3)
        data.add_node_filter(f4)

        will result in elements for which
        
        (f1(element) and f2(element) and f3(element)) or f4(element)

        is False getting filtered out and elements for which that is True being kept
        """
        OSMData.add_filter(self.node_filters, *filters)
    
    def add_way_filter(self, *filters):
        """
        Adds filters.
        Filters are functions that take a ElementTree element and this OSMData instance
        as arguments and returns a truthful value if said element should be
        kept.
        Filters given in the same method call are joined with "and",
        while filters given in subsequent calls are joined with "or".
        For example

        data = OSMData()
        data.add_way_filter(f1, f2, f3)
        data.add_way_filter(f4)

        will result in elements for which
        
        (f1(element) and f2(element) and f3(element)) or f4(element)

        is False getting filtered out and elements for which that is True being kept
        """
        OSMData.add_filter(self.way_filters, *filters)
    
    @classmethod
    def add_filter(cls, filterlist, *filters):
        filterlist.append(filters)

    def do_filter(self):
        self.filter(self.included_nodes, self.nodes, self.node_filters)
        self.filter(self.included_ways, self.ways, self.way_filters)

    def filter(self, idset, elemdict, filterlist):
        filterset = set()
        for elemid in idset:
            elem = elemdict[elemid]
            ok = False
            for filt in filterlist:
                current_ok = True
                for f in filt:
                    if not f(elem, self):
                        current_ok = False
                        break
                if current_ok:
                    ok = True
                    break
            if not ok:
                filterset.add(elemid)
        idset -= filterset


    def prepare_for_save(self):
        root = self.tree.getroot()
        for way in self.ways.values():
            wayid = OSMData.get_elem_id(way)
            # Remove dropped ways
            if wayid not in self.included_ways:
                root.remove(way)
            else:
                for child in way:
                    # Add nodes that belong to the way even if they were dropped by some other filter
                    if child.tag == OSMData.TAG_WAY_NODE:
                        self.included_nodes.add(int(child.get(OSMData.ATTRIB_REF)))
        for node in self.nodes.values():
            nodeid = OSMData.get_elem_id(node)
            # Remove dropped nodes
            if nodeid not in self.included_nodes:
                root.remove(node)

    def save(self, path):
        self.tree.write(path, encoding='utf-8', xml_declaration=True)
    
def areaFilter(elem, osmdata):
    """
    Currently filters in only areas with "landuse" in the accepted landuses list.
    """
    return OSMData.get_tag(elem, OSMData.KEY_LANDUSE) in OSMData.ACCEPTED_LANDUSES

def trailFilter(elem, osmdata):
    """
    Currently filters in all trails with "highway" in the accepted highways list.
    """
    return OSMData.get_tag(elem, OSMData.KEY_HIGHWAY) in OSMData.ACCEPTED_HIGHWAYS

class WayCoordinateFilter:
    def __init__(self, minx, maxx, miny, maxy):
        self.minx = minx
        self.miny = miny
        self.maxx = maxx
        self.maxy = maxy
    def filter(self, elem, osmdata):
        for ref in elem.iter(OSMData.TAG_WAY_NODE):
            try:
                node = osmdata.nodes[int(ref.get(OSMData.ATTRIB_REF))]
                x = float(node.get(OSMData.ATTRIB_LON))
                y = float(node.get(OSMData.ATTRIB_LAT))
            except ValueError:
                continue # Just skip any dirty data
            if self.minx <= x and x <= self.maxx and self.miny <= y and y <= self.maxy:
                return True
        return False

def merge(osm_datas):
    """
    Merges several OSMdatas into one. Takes a list of OSMdatas as a parameter. 
    Returns a combined OSMdata. Entity doubles are not checked or removed, so the 
    combined XML may combine doubles. The combined XML tree is the attribute tree 
    of the OSMdata that is returned.
    """
    trees = []
    for osm_data in osm_datas:
        trees.append(osm_data.tree)
    result_xml_element_tree = None
    for osm_tree in trees:
        if result_xml_element_tree is None:
            result_xml_element_tree = osm_tree
        else:
            for child in osm_tree.getroot().iterfind('.'):
                result_xml_element_tree.getroot().extend(child) 
    if result_xml_element_tree is not None:
        resultOSMdata = OSMData()
        resultOSMdata.tree = result_xml_element_tree
        return resultOSMdata