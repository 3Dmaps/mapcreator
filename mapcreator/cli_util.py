import click
import sys, traceback
from mapcreator import building
from mapcreator import persistence
from mapcreator import echoes
from mapcreator.state import FileAddResult


"""
The x_or_error functions return a falsy value if an error was thrown.
If everything went OK, they return either True or resulting state, depending on the function.
"""

def init_or_error():
    try:
        state = persistence.init_state()
    except Exception as e:
        echoes.error('Unable to initialize project: {}'.format(e))
        return None
    else:
        return state

def load_or_error():
    try:
        state = persistence.load_state()
    except Exception as e:
        echoes.error('Unable to load or initialize the project: {}'.format(e))
        return None
    else:
        return state

def save_or_error(state):
    try:
        persistence.save_state(state)
    except Exception as e:
        echoes.error('Unable to save changes to the project! No changes done!')
        echoes.error('What went wrong: {}'.format(e))
        return None
    else:
        return state

def clear_or_error():
    try:
        persistence.clear_state()
    except Exception as e:
        echoes.error('Unable to reset project: {}'.format(e))
        return False
    else:
        return True

def build_init_or_error():
    try:
        building.init_build()
    except Exception as e:
        echoes.error('Unable to initialize build: {}'.format(e))
        return False
    else:
        return True

def build_clean_or_error():
    try:
        building.cleanup()
    except Exception as e:
        echoes.error('Unable to clean temporary build files: {}'.format(e))
        return False
    else: 
        return True

def add_files(files, add_method_name):
    state = load_or_error()
    if not state: return
    echoes.info('Adding files to project...')
    count = 0
    for fpath in files:
        result = getattr(state, add_method_name)(fpath) # Call method whose name is add_method_name
        if result == FileAddResult.DOESNT_EXIST:
            echoes.error('File "{}" doesn\'t exist!'.format(fpath))
        elif result == FileAddResult.ALREADY_ADDED:
            echoes.warn('{} has already been added to this project'.format(fpath))
        elif result == FileAddResult.SUCCESS:
            echoes.info('"{}" added'.format(fpath))
            count += 1
        else:
            echoes.error('Unrecognized FileAddResult {} when trying to add {}!'.format(result, fpath))
    if count > 0:
        if not save_or_error(state): return
        if count == len(files):
            echoes.success("{} files added to the project successfully!".format(len(files)))
        else:
            echoes.warn("{} files (out of {}) added to the project successfully".format(count, len(files)))
    else:
        echoes.warn('No files were added.')

def clear_files(clear_method_name, files_type):
    state = load_or_error()
    if not state: return
    echoes.info('Clearing {} files...'.format(files_type))
    getattr(state, clear_method_name)() # Call method whose name is clear_method_name
    if save_or_error(state):
        echoes.success('All {} files cleared successfully!'.format(files_type))
    

def parse_color(line, debug = False):
    PARSER = [str, int, int, int]
    parts = line.split(' ')
    if len(parts) != len(PARSER):
        echoes.error("Invalid number of arguments on a line!")
        echoes.error("(Should be {}, was {})".format(len(PARSER), len(parts)))
        if debug: echoes.info("Line was: {}".format(line))
        return None
    try:
        for i, func in enumerate(PARSER):
            parts[i] = func(parts[i])
    except ValueError as e:
        echoes.error("Invalid color format!")
        echoes.error("({})".format(e))
        if debug: echoes.info("Line was: {}".format(line))
        return None
    return parts

def validate_color(r, g, b):
    for value in (r, g, b):
        if value < 0 or value > 255:
            echoes.error("Invalid color value {}!".format(value))
            echoes.info("(Should be between 0 and 255)")
            return False
    return True

def validate_resolution(reso, lower, upper):
    if reso < lower or reso > upper:
        echoes.error("Invalid resolution {}!".format(reso))
        echoes.info("(Should be between {} and {})".format(lower, upper))
        return False
    return True

def do_build(files, statusclass, actions, state, debug = False):
    has_errors = False
    outfiles = []
    echoes.info('Processing...')
    buildstatus = statusclass(0, files, state)
    errors = []
    with click.progressbar(actions, bar_template=echoes.PROGRESS_BAR_TEMPLATE, show_eta=False) as bar:
        for action in bar:
            try:
                action(buildstatus, debug)
            except Exception as e:
                errors.append(e)
                if debug:
                    e_type, e_value, e_traceback = sys.exc_info()
                    errors.extend(traceback.format_tb(e_traceback))
                has_errors = True
    if errors:
        echoes.error('Exceptions caught:')
        for e in errors:
            echoes.error(e)
    for line in str(buildstatus).split('\n'):
        echoes.info(line)
    outfiles.extend(buildstatus.get_result_files())
    return (outfiles, has_errors)