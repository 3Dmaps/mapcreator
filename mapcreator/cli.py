import click
import subprocess
from os import path
from mapcreator import building
from mapcreator import persistence
from mapcreator.cli_util import *
from mapcreator.echoes import *
from mapcreator.state import FileAddResult


@click.group()
def cli():
    """
    An utility for creating correctly formatted data for the 3Dmaps application.

    To get more information of what some command does, try
    mapcreator [command] --help

    Part of Software production project course, Spring 2018, University of Helsinki.
    """
    pass

@click.command()
def hello():
    """Says 'Hello world!', very successfully!"""
    success('Hello world!')

@click.command()
@click.option('--force', '-f', is_flag=True, help='Force initialization, even if project already exists')
def init(force):
    """
    Initializes the project. 
    A project is also automatically initialized by any command that needs persistence,
    so calling init is just a more verbose way to do it.
    """
    if persistence.state_exists() and not force:
        warn('Project already exists, no initialization done.')
        info('If you want to reset the project and initialize a new project, do:')
        info('mapcreator init --force')
        return
    info('Initializing project...')
    if init_or_error():
        success('Project initialized!')

@click.command()
@click.argument('files', nargs=-1)
def add_height_files(files):
    """Adds given height data files to the project"""
    if len(files) == 0:
        warn('No files were specified.')
        info('Try mapcreator add_height_files [file 1] [file 2] ... [file n]')
        return
    add_files(files, 'add_height_file')

@click.command()
@click.argument('files', nargs=-1)
def add_osm_files(files):
    """Adds open street map files to the project"""
    if len(files) == 0:
        warn('No files were specified.')
        info('Try mapcreator add_osm_files [file 1] [file 2] ... [file n]')
        return
    add_files(files, 'add_osm_file')

@click.command()
@click.argument('ulx', type=float)
@click.argument('uly', type=float)
@click.argument('lrx', type=float)
@click.argument('lry', type=float)
def set_window(ulx, uly, lrx, lry):
    """
    Specifies projection subwindow. 
    Uses upper left (ulx, uly) and lower right (lrx, lry) corners.

    If any of the coordinates is negative, you'll need to seperate the arguments from the command
    with a double dash; for example
    
    mapcreator set_window -- 110.3 -13.2 110.5 -12.1
    """
    state = load_or_error()
    if not state: return
    info('Setting window to {} -> {}'.format((ulx, uly), (lrx, lry)))
    state.set_window(ulx, uly, lrx, lry)
    if save_or_error(state):
        success('Window set to {} -> {}'.format((ulx, uly), (lrx, lry)))

@click.command()
@click.option('--output', '-o', default='3dmapdata.zip', help='Output file name. Default: 3dmapdata.zip')
@click.option('--force', '-f', is_flag=True, help='Build even if output file already exists')
@click.option('--debug', '-d', is_flag=True, help='Causes debug information to be printed during the build')
@click.option('--clean/--no-clean', default=True, help='Specifies whether to clean temporary build files after building')
def build(output, force, debug, clean):
    """
    Builds the project.
    Transforms and translates all output files to format used by the 3DMaps-application and packages them for easy transportation.
    """

    if path.exists(output) and not force:
        error('File {} already exists!'.format(output))
        info('If you wish to build and overwrite this file, do:')
        info('mapcreator build --force')
        return

    if not persistence.state_exists(): 
        warn('No project found in current working directory.')
        return
    state = load_or_error()
    if not state: return
    if not state.has_height_files():
        error('No height files have been added to the current project! There\'s nothing to build!')
        return
    
    if not build_init_or_error(): return

    highlight('STARTING BUILD')
    info('Output file is {}'.format(output))

    outfiles = []
    has_errors = False
    current_height_files = combine_files(state.height_files)
    state.set_height_files(current_height_files)
    #info('Length of height files: ' + str(len(current_height_files)))
    build_outfiles, build_has_errors = do_build(
        state.height_files, building.BuildStatus, building.BUILD_ACTIONS, state, debug
        )
    has_errors |= build_has_errors
    outfiles.extend(build_outfiles)

    osm_outfiles, osm_has_errors = do_build(
        state.osm_files, building.OSMStatus, building.OSM_ACTIONS, state, debug
    )
    has_errors |= osm_has_errors
    outfiles.extend(osm_outfiles)

    info('Building package...')
    try:
        building.package(output, outfiles)
    except Exception as e:
        error('Unable to create package: {}'.format(e))
        has_errors = True
    if clean:
        info('Cleaning up...')
        if not build_clean_or_error():
            has_errors = True
    else:
        info('Not cleaning temporary build files')
    if has_errors:
        warn('Build done (but there were errors)')
    else:
        success('BUILD SUCCESSFUL!')

@click.command()
def clean_temp_files():
    """Cleans up temporary build files"""
    if not building.temp_build_files_exist():
        info('No temporary build files to clean up.')
        return
    info('Cleaning up temporary build files...')
    if build_clean_or_error():
        success('Cleaned up!')

@click.command()
def status():
    """Shows the status of the current project"""
    if persistence.state_exists():
        highlight('CURRENT PROJECT STATE')
        info('-Persistence location: {}'.format(path.abspath(persistence.state_path())))
        state = persistence.load_state()
        for line in str(state).split('\n'):
            info(line)
    else:
        info('No project found in current working directory.')

@click.command()
def reset():
    """Clears the current project."""
    if not persistence.state_exists():
        info("No project found in current working directory. No need to reset!")
        return
    info('Resetting project...')
    if clear_or_error():
        success('Project has been reset.')

cli.add_command(hello)
cli.add_command(init)
cli.add_command(add_height_files)
cli.add_command(add_osm_files)
cli.add_command(set_window)
cli.add_command(status)
cli.add_command(reset)
cli.add_command(build)
cli.add_command(clean_temp_files)
