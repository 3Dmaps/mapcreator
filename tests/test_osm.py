import os
import shutil
from mapcreator.osm import OSMData, areaFilter, WayCoordinateFilter, trailFilter, OSMMerger
from mapcreator.building import OSMStatus
from os import path, mkdir
from util import get_resource_path, assert_xml_equal

CLEAN_TEMP_DIR = False
TEMP_DIR = '.test_osm'

def setup_function(function):
    if not path.exists(TEMP_DIR):
        mkdir(TEMP_DIR)

def test_osmdata():
    def node_filter_a(element, data):
        return element.get('user') == 'zmeu'
    def node_filter_b(element, data):
        for tag in element:
            if tag.get('k') == 'zmeuicon':
                return True
        return False
    def way_filter_a(element, data):
        for tag in element.iter('tag'):
            if tag.get('k') == 'zmeucolor' and tag.get('v') == 'yellow':
                return True
        return False
    def way_filter_b(element, data):
        for tag in element.iter('tag'):
            if tag.get('k') == 'highway' and tag.get('v') == 'footway':
                return True
        return False
    def way_filter_c(element, data):
        for tag in element.iter('tag'):
            if tag.get('k') == 'EXPORT_ROUTE' and tag.get('v') == 'true':
                return True
        return False
    data = OSMData.load(get_resource_path('test_osm_input.xml'))
    data.add_node_filter(node_filter_a) #Node filter a
    data.add_node_filter(node_filter_b) #Or node filter b
    data.add_way_filter(way_filter_a, way_filter_b) #Way filter a and way filter b
    data.add_way_filter(way_filter_c) #Or way filter c
    result_path = path.join(TEMP_DIR, 'test_osm_result.xml')
    data.do_filter()
    data.prepare_for_save()
    data.save(result_path)
    assert_xml_equal(get_resource_path('test_osm_expected.xml'), result_path)

def test_osm_terrainfilter():
    data = OSMData.load(get_resource_path('test_osm_terrains_input.xml'))
    data.add_way_filter(areaFilter)
    
    result_path = path.join(TEMP_DIR, 'test_osm_terrains_result.xml')
    data.do_filter()
    data.prepare_for_save()
    data.save(result_path)
    assert_xml_equal(get_resource_path('test_osm_terrains_expected.xml'), result_path)

def test_way_coordinate_filter():
    wcf = WayCoordinateFilter(-112.060, -112.000, 36.050, 36.109)
    data = OSMData.load(get_resource_path('test_osm_input.xml'))
    data.add_way_filter(wcf.filter)
    result_path = path.join(TEMP_DIR, 'test_osm_result.xml')
    data.do_filter()
    data.prepare_for_save()
    data.save(result_path)
    assert_xml_equal(get_resource_path('test_osm_expected_way_coord_filter.xml'), result_path)

def test_osm_trailfilter():
    data = OSMData.load(get_resource_path('test_osm_trails_input.xml'))
    data.add_way_filter(trailFilter)
    
    result_path = path.join(TEMP_DIR, 'test_osm_trails_result.xml')
    data.do_filter()
    data.prepare_for_save()
    data.save(result_path)
    assert_xml_equal(get_resource_path('test_osm_trails_expected.xml'), result_path)

def test_osm_trailfilter_with_coordinates():
    wcf = WayCoordinateFilter(-113.0, -112.0, 36.0, 37.0)
    data = OSMData.load(get_resource_path('test_osm_trails_input.xml'))
    data.add_way_filter(trailFilter, wcf.filter)
    
    result_path = path.join(TEMP_DIR, 'test_osm_trails_and_coordinates_result.xml')
    data.do_filter()
    data.prepare_for_save()
    data.save(result_path)
    assert_xml_equal(get_resource_path('test_osm_trails_and_coordinates_expected.xml'), result_path)

def test_osm_merger():
    trails = OSMData.load(get_resource_path('test_osm_trails_input.xml'))
    terrains = OSMData.load(get_resource_path('test_osm_terrains_input.xml'))
    combined = OSMMerger().merge([trails, terrains])
    result_path = path.join(TEMP_DIR, 'test_osm_combined_result.xml')
    combined.tree.write(result_path, encoding='utf-8', xml_declaration=True)
    assert_xml_equal(get_resource_path('test_osm_combined_result.xml'), result_path)
    
def teardown_function(function):
    if CLEAN_TEMP_DIR and path.exists(TEMP_DIR):
        shutil.rmtree(TEMP_DIR)

    

