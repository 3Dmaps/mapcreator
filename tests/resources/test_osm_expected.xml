<?xml version="1.0" encoding="utf-8"?>
<osm xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema" version="0.6">
  <node timestamp="2017-11-23T15:46:19Z" uid="1" user="zmeu" visible="true" version="1" changeset="1" id="111" lat="35.9432567" lon="-112.1851486">
    <tag k="ele" v="2007.46655273438" />
    <tag k="name" v="Apex" />
    <tag k="is_in" v="Coconino,Arizona,Ariz.,AZ,USA" />
    <tag k="place" v="hamlet" />
    <tag k="gnis:id" v="25248" />
    <tag k="gnis:Class" v="Populated Place" />
    <tag k="gnis:County" v="Coconino" />
    <tag k="gnis:ST_num" v="04" />
    <tag k="import_uuid" v="bb7269ee-502a-5391-8056-e3ce0e66489c" />
    <tag k="gnis:ST_alpha" v="AZ" />
    <tag k="gnis:County_num" v="005" />
    <tag k="EXPORT_POI" v="true" />
    <tag k="zmeuicon" v="city" />
  </node>
  <node timestamp="2017-11-23T15:46:19Z" uid="1" user="not_zmeu" visible="true" version="1" changeset="1" id="999" lat="36.2301946" lon="-111.9738232">
    <tag k="ele" v="2595.87329101562" />
    <tag k="name" v="Vista Encantada" />
    <tag k="gnis:id" v="24675" />
    <tag k="tourism" v="viewpoint" />
    <tag k="gnis:Class" v="Populated Place" />
    <tag k="gnis:County" v="Coconino" />
    <tag k="gnis:ST_num" v="04" />
    <tag k="import_uuid" v="bb7269ee-502a-5391-8056-e3ce0e66489c" />
    <tag k="gnis:ST_alpha" v="AZ" />
    <tag k="gnis:County_num" v="005" />
    <tag k="EXPORT_POI" v="true" />
    <tag k="zmeuicon" v="viewpoint" />
  </node>
  <node timestamp="2017-11-23T15:46:19Z" uid="1" user="zmeu" visible="true" version="1" changeset="1" id="112" lat="36.0176913" lon="-112.1253312" />
  <node timestamp="2017-11-23T15:46:19Z" uid="1" user="zmeu" visible="true" version="1" changeset="1" id="113" lat="36.0537344" lon="-112.1439038" />
  <node timestamp="2017-11-23T15:46:19Z" uid="1" user="not_zmeu" visible="true" version="1" changeset="1" id="114" lat="36.0537344" lon="-112.1439038" />
  <node timestamp="2017-11-23T15:46:19Z" uid="1" user="zmeu" visible="true" version="1" changeset="1" id="117" lat="36.1011288" lon="-112.0608877" />
  <node timestamp="2017-11-23T15:46:19Z" uid="1" user="zmeu" visible="true" version="1" changeset="1" id="118" lat="36.104251" lon="-112.0378661" />
  <node timestamp="2017-11-23T15:46:19Z" uid="1" user="zmeu" visible="true" version="1" changeset="1" id="119" lat="36.1057122" lon="-112.0359378" />
  <node timestamp="2017-11-23T15:46:19Z" uid="1" user="zmeu" visible="true" version="1" changeset="1" id="120" lat="36.1082665" lon="-112.0233688" />
  <node timestamp="2017-11-23T15:46:19Z" uid="1" user="random" visible="true" version="1" changeset="1" id="121" lat="36.1082433" lon="-112.0910089" />
  <node timestamp="2017-11-23T15:46:19Z" uid="1" user="zmeu" visible="true" version="1" changeset="1" id="123" lat="36.1082665" lon="-112.0233688" />
  <way timestamp="2017-11-23T15:46:19Z" uid="1" user="zmeu" visible="true" version="1" changeset="1" id="200">
    <nd ref="111" />
    <nd ref="112" />
    <nd ref="113" />
    <nd ref="114" />
    <tag k="name" v="Powell Point Trail" />
    <tag k="highway" v="footway" />
    <tag k="EXPORT_ROUTE" v="false" />
    <tag k="zmeucolor" v="yellow" />
    <tag k="zmeuname" v="Powell Point Trail" />
    <tag k="intermediate_points" v="1856359973,Powell Point,IntermediatePoint_TRAILHEAD,,1856359973,Powell Point,IntermediatePoint_TRAILHEAD,,1856359961,Powell Point,IntermediatePoint_TRAILHEAD," />
  </way>
  <way timestamp="2017-11-23T15:46:19Z" uid="1" user="zmeu" visible="true" version="1" changeset="1" id="128437048">
    <nd ref="117" />
    <nd ref="118" />
    <nd ref="119" />
    <tag k="foot" v="designated" />
    <tag k="highway" v="path" />
    <tag k="lit" v="no" />
    <tag k="name" v="Cape Royal Trail" />
    <tag k="EXPORT_ROUTE" v="true" />
    <tag k="zmeucolor" v="violet" />
    <tag k="zmeuname" v="Cape Royal Trail" />
    <tag k="intermediate_points" v="1419410942,Trailhead,IntermediatePoint_TRAILHEAD,,1414803509,Cape Royal,IntermediatePoint_TRAILHEAD," />
  </way>
  <way timestamp="2017-11-23T15:46:36Z" uid="1" user="zmeu" visible="true" version="1" changeset="1" id="100000000096">
    <nd ref="120" />
    <nd ref="121" />
    <tag k="highway" v="footway" />
    <tag k="source" v="Steiner tree path" />
    <tag k="EXPORT_ROUTE" v="false" />
    <tag k="zmeucolor" v="yellow" />
  </way>
</osm>